DOCS-6919-Improve selectors documentation (#1090)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dataweave-selectors.adoc

> a variable, an object literal, an array literal, or the invocation of a

Reviewer: Instructions for edit:

None. Default copyedit of green changed text.

|===

[NOTE]
====

> The examples below use these selectors. For additional examples, see

Reviewer: below > following
GLOBAL, per style ref.
```suggestion
The following examples use these selectors. For additional examples, see
```
Will it be clear to your audience what "these selectors" refers to?

xref:dataweave-cookbook-extract-data#[Extract Data].
====

== Rules for Matching

In DataWeave, a name that matches the key of an object (such as the key
`"someName"` in the JSON object `{ "someName" : "Data Weave" }`) can select
data from that object. For a simple example, see <<single_value_selector>>.

== Dynamic Selector

You perform dynamic selection with a `nameExpression` within
square brackets (`[``]`) that returns a `String` type.

* Single Value: `payload[(nameExpression)]`
* Multi Value: `payload[*(nameExpression)]`
* Attribute: `payload[@(nameExpression)]`
* Key Value: `payload[&(nameExpression)]`
* Single Value with namespace: `payload.ns0#"$(nameExpression)"`

=== Example 
This example shows how to dynamically select a single value. It uses the value on the "ref" property to select the value with that key.

=== Input

[source,JSON,linenums]
----
{ "ref": "name", "name": "Data Weave" }
----

> ==== Source

Reviewer: Does this really require a Level-4 heading?

[source,DataWeave,linenums]
----
%dw 2.0
output application/json
---
payload[(payload.ref)]
----

> ==== Output

Reviewer: Should this be at the same level as ===Input?
...or should Input really be ====Input, subordinate to ===Example?
Yikes, if we can figure out how to reduce this to three levels, that would be great.

[source,JSON,linenums]
----
"Data Weave"
----

=== Example 
This example shows how to dynamically select a single value with namesapce. It uses the value on the "ref" property to select the value with that key.

> === Input

Reviewer: ditto, GLOBAL: ====Input?

[source,xml,linenums]
----
<root ref="table">
<f:table xmlns:f="https://www.w3schools.com/furniture">Manzana</f:table>
    <h:table xmlns:h="http://www.w3.org/TR/html4/">Banana</h:table>    
</root>
----

==== Source

[source,DataWeave,linenums]
----
%dw 2.0
output application/json
ns h http://www.w3.org/TR/html4/
---
payload.root.h#"$(payload.root.@ref)"
----

==== Output

[source,JSON,linenums]
----
"Banana"
----

[[single_value_selector]]
== Single-Value Selector

> This selector matches the value of the first key-value pair where the key

Reviewer: ```suggestion
This selector matches the value of the first key-value pair in which the key
```

matches the given selector name. The selector can be applied to an Object or
to an Array. On an Array, the selector applies to all Object values inside the

> array and ignores all values that are not Object values. If no key matches,

Reviewer: I suggest lowercasing both "object" and "array."

Reviewer 2: we're saying it's an value of the Array type, emphasizing the type here

> the value `null` returns.

Reviewer: ```suggestion
the value `null` is returned.
```

=== Example

This example shows how the selector works on a simple JSON object.

==== Input

The input payload is an object with the key "name" and the

> value "DataWeave".

Reviewer: ```suggestion
value "DataWeave":
```

[source,JSON,linenums]
----
{ "name": "Data Weave" }
----

==== Source

The DataWeave script uses `payload.name` to select the
value of the object that has the key `"name"`.
Note that `payload.name` and `payload."name"` are both valid
ways to perform the selection.

[source,DataWeave,linenums]
----
%dw 2.0
output application/json
---
payload.name
----

==== Output

> The script outputs the String value of the input object, "Data Weave".

Reviewer: ```suggestion
The script outputs the String value of the input object, "Data Weave":
```

[source,JSON,linenums]
----
"Data Weave"
----

=== Example

> On an Array, the selector applies to every element.

Reviewer: array

==== Input

The input payload is an array that contains two objects with the same

> key, "name".

Reviewer: ```suggestion
key, "name":
```

[source,JSON,linenums]
----
[
  {
    "name": "Arg"
  },
  {
    "name": "Japan"
  }
]
----

==== Source

The DataWeave script uses `payload.name` to select the

> value of any objects in input payload that have the key, "name".

Reviewer: ```suggestion
value of any objects in input payload that have the key "name":
```

[source,DataWeave,linenums]
----
%dw 2.0
output application/json
---
payload.name
----

==== Output

The script outputs an array that contains the values of the input objects

> with the key "name": "Arg" and "Japan".

Reviewer: ```suggestion
with the key "name": "Arg" and "Japan":
```

[source,json, linenums]
----
["Arg", "Japan"]
----

[[namespace_example1]]
== Namespace Selector

The DataWeave namespace selector supports the use of a name that matches a

> namespace or a local part. To match a namespace, the name must match the

Reviewer: Does your audience know what "a local part" is? Because I don't...lol....

Reviewer 2: there's an example. It's a common term.

Reviewer 2: it's explained later also

namespace and the local part. If the name only matches the local part, the

> selector will match all namespaces with that local part, regardless of the

Reviewer: present tense GLOBAL
```suggestion
selector matches all namespaces with that local part, regardless of the
```

namespace.

Note that XML namespaces and their prefixes are defined in the `xmlns` attribute
of an XML element. For example, the element
`<h:table xmlns:h="http://www.w3.org/TR/html4/"/>` defines a namespace that
assigns the prefix `h` to the namespace  `http://www.w3.org/TR/html4/`.
All elements in that namespace contain the `h` prefix. For example, in the
element `<h:table/>`, `h` is the prefix for the namespace, and `table` is the
local part.

=== Example

This example selects values from XML elements using the namespace and local part.

==== Input

The input payload contains XML elements, `h:table` and `f:table`, that have

> different namespaces but the same local names (`table`).

Reviewer: local names or local part names?

Reviewer: ```suggestion
different namespaces but the same local names (`table`):
```

[source,xml, linenums]
----
<root>

  <h:table xmlns:h="http://www.w3.org/TR/html4/">
    <h:tr>
      <h:td>Apples</h:td>
      <h:td>Bananas</h:td>
    </h:tr>
  </h:table>

  <f:table xmlns:f="https://www.w3schools.com/furniture">
    <f:tr>
      <f:name>African Coffee Table</f:name>
      <f:width>80</f:width>
      <f:length>120</f:length>
    </f:tr>
  </f:table>

</root>
----

[[namespace_variable]]
==== Source

The DataWeave script selects XML content from specified namespaces. The
script's header creates namespace (`ns`) variables `html` and `furniture`
to store namespaces from the input payload. To select the children of `h:table`

> into element `a` and the children of `f:table` elements into element `b`, the

Reviewer: Do we need to use "select into" as a verb? Is this "move into"? Or "select and move into"?

Reviewer 2: we select the value of `payload.root.html#table` into the an object with the key `a`, for example.

script uses the `html` and `furniture` namespace variables with a

> namespace selector (`#`) and local part (`table`).

Reviewer: ```suggestion
namespace selector (`#`) and local part (`table`):
```

[source,DataWeave, linenums]
----
%dw 2.0
output application/xml
ns html http://www.w3.org/TR/html4/
ns furniture https://www.w3schools.com/furniture
---
root: {
     a: payload.root.html#table,
     b: payload.root.furniture#table
 }
----

==== Output

The script outputs children of the `h:table` element under element `a` and

> the children of the `f:table` under element `b`.

Reviewer: ```suggestion
the children of the `f:table` under element `b`:
```

[source,xml, linenums]
----
<?xml version='1.0' encoding='UTF-8'?>
<root>
  <a>
    <h:tr xmlns:h="http://www.w3.org/TR/html4/">
      <h:td>Apples</h:td>
      <h:td>Bananas</h:td>
    </h:tr>
  </a>
  <b>
    <f:tr xmlns:f="https://www.w3schools.com/furniture">
      <f:name>African Coffee Table</f:name>
      <f:width>80</f:width>
      <f:length>120</f:length>
    </f:tr>
  </b>
</root>
----

> === Example

Reviewer: I wonder if we can reword these repeated ===Example titles so that they are specific to their use cases....

This example selects XML values using a local name `table` from the `h:table`
element.

==== Input

The input payload contains XML with a `table` element that contains

> the namespace `"http://www.w3.org/TR/html4/"`.

Reviewer: ```suggestion
the namespace `"http://www.w3.org/TR/html4/"`:
```

[source,xml, linenums]
----
<root>
  <h:table xmlns:h="http://www.w3.org/TR/html4/">
    <h:tr>
      <h:td>Apples</h:td>
      <h:td>Bananas</h:td>
    </h:tr>
  </h:table>
</root>
----

==== Source

The DataWeave script selects values of children in the `table` element

> by using a local name of the XML element `table` instead of its namespace.

Reviewer: ```suggestion
by using a local name of the XML element `table` instead of its namespace:
```

[source,DataWeave, linenums]
----
%dw 2.0
output application/xml
---
root: { a: payload.root.table }
----

==== Output

The script outputs XML that contains the values of children in the

> `table` element.

Reviewer: ```suggestion
`table` element:
```

[source,xml, linenums]
----
<?xml version='1.0' encoding='UTF-8'?>
<root>
  <a>
    <h:tr xmlns:h="http://www.w3.org/TR/html4/">
      <h:td>Apples</h:td>
      <h:td>Bananas</h:td>
    </h:tr>
  </a>
</root>
----

[[attribute_selector]]
== Attribute Selector

The attribute selector returns the first attribute value that matches the
selected name expression. If no key matches, the selection returns the value
`null`.

> 

Reviewer: Missing ===Example heading...although I kind of like it this way instead of the way we have been doing it, because we omit that level of heading and can promote the subs up a level....

The following example selects the value of an XML attribute.

=== Input

The input payload is an XML `user` element that contains the attribute

> `name="Weave"`.

Reviewer: ```suggestion
`name="Weave"`:
```
GLOBALLY change the period ending the sentence before the syntax example into a colon. I'll stop marking it.

[source,xml, linenums]
----
<user name="Weave"/>
----

=== Source

The DataWeave script selects the value of the `name` attribute
from the input payload (`<user name="Weave"/>`). Notice that the `@`
indicates an attribute selection.

[source,DataWeave, linenums]
----
%dw 2.0
output application/json
---
payload.user.@name
----

=== Output

The DataWeave script outputs the value of `name` attribute.

[source,json,linenums]
----
"Weave"
----

[[multi_value_selector]]
== Multi-Value Selector

Instead of returning the value of the first matching key in an array
of objects, the multi-value selector (`*`) returns an array containing
values to the matching key (for example, `*user` where `user` is

> the key). The selector does not return the values of descendants,

Reviewer: Should we use "children" here, to be consistent with previous text?
```suggestion
the key). The selector does not return the values of children; it returns  
```

Reviewer 2: It's all descendants, not just the children. Includes grandchildren, etc.

Reviewer 2: replied on slack

> only those at the specified level. If no key matches, the selection

Reviewer: ```suggestion
only those values at the specified level. If no key matches, the selection
```

Reviewer 2: keeping values

returns the value `null`.

=== Input

The input payload contains an XML array of `user` elements.

[source,xml, linenums]
----
<users>
  <user>Weave</user>
  <user>BAT</user>
  <user>TF</user>
</users>
----

=== Source

> The DataWeave script uses the `*user` to select the values of

Reviewer: ```suggestion
The DataWeave script uses `*user` to select the values of
```

all `user` elements from the input payload.

[source,DataWeave, linenums]
----
%dw 2.0
output application/json
---
payload.users.*user
----

=== Output

The script outputs an array of `user` values.

[source,json,linenums]
----
[
  "Weave",
  "BAT",
  "TF"
]
----

[[descendant_selector]]
== Descendant Selector

> The descendant selector returns a list of all children. You can directly chain

Reviewer: okay...I guess we have to use both "children" and "descendant"....

this selector to any other selector without using a single `.`. For example,

> `payload..` recursively returns an array of all the child values and the values

Reviewer: ```suggestion
`payload..` recursively returns an array of all the child values, the values
```

> of their children, and so on. You can also chain the selector to another element

Reviewer: ```suggestion
of _their_ children, and so on. You can also chain the selector to another element
```

|===

== Rules for matching 

> All selectors that uses a `name` to select are composed of a `Namespace` and a `Local Part`. When using namespace it is only going to match a `Key` with that `Namespace` and that `Local Part`. If the `name` only has a `Local Part` it is going to match all the `Key` whoes `Local Part` are equal dispaite it has `Namespace` or not.

Reviewer 2: This needs some clarification or an example.

== Dynamic Selector

The way to do dynamic selection is by using the `[``]` way of selecting. 

> - Single Value: `payload[(nameExpression)]`

Reviewer 2: Need examples for all Single through Key Value.

==== Input

[source,json, linenums]
----
[{
    "name": "Arg"
},
{
    "name": "Japan"
}
]
----

==== Source

[source,DataWeave, linenums]
----
%dw 2.0
output application/json
---
payload.name
----

==== Output

[source,json, linenums]
----
["Arg", "Japan"]
----

=== Example

Shows how to select an xml with namespace or directly with local name

==== Input

[source,xml, linenums]
----
<root>

<h:table xmlns:h="http://www.w3.org/TR/html4/">
  <h:tr>
    <h:td>Apples</h:td>
    <h:td>Bananas</h:td>
  </h:tr>
</h:table>

<f:table xmlns:f="https://www.w3schools.com/furniture">
  <f:name>African Coffee Table</f:name>
  <f:width>80</f:width>
  <f:length>120</f:length>
</f:table>

</root>
----

==== Source

[source,DataWeave, linenums]
----
%dw 2.0
output application/xml
ns html http://www.w3.org/TR/html4/
ns furniture https://www.w3schools.com/furniture
---
root: {
     a: payload.root.html#table,
     b: payload.root.furniture#table
 }    
----

==== Output

[source,xml, linenums]
----
<?xml version='1.0' encoding='UTF-8'?>
<root>
  <a>
    <h:tr xmlns:h="http://www.w3.org/TR/html4/">
      <h:td>Apples</h:td>
      <h:td>Bananas</h:td>
    </h:tr>
  </a>
  <b>

>     <f:name xmlns:f="https://www.w3schools.com/furniture">African Coffee Table</f:name>

Reviewer 2: It is weird that the output adds the namespace attribute to all `f:`-prefixed elements but not to all the `h:` elements. Is there a reason for this behavior.

      <h:td>Apples</h:td>
      <h:td>Bananas</h:td>
    </h:tr>
  </a>  
</root>
----

== Attribute Selector

The attribute selector returns the first Attribute value that matches the selected name expression. If no key was matched it return `null`

=== Example

Shows how to select an xml attribute

==== Input

[source,xml, linenums]
----
<user name="Weave"/>
----

==== Source

[source,DataWeave, linenums]
----
%dw 2.0
output application/json
---
payload.user.@name   
----

==== Output

[source,json,linenums]
----
"Weave"
----

== Multi Value Selector

Returns an Array of all the matched keys value. If no key was matched it return `null`

=== Example

Shows how to select multiple values

==== Input

[source,xml, linenums]
----
<users>
    <user>Weave</user>
    <user>BAT</user>
    <user>TF</user>
</users>
----

==== Source

[source,DataWeave, linenums]
----
%dw 2.0
output application/json
---
payload.users.*user
----

==== Output

[source,json,linenums]
----
[
  "Weave",
  "BAT",
  "TF"
]
----

== Descendant Selector

Returns the list of all children. This selector can be chained with any other selector directly without requiering any `.`. For example

> `payload..` returns all the child values and the value of its children and recursively. But it can also be chained with `payload..name` where it will select the name to all its descendants, or `payload..*name` where it will select all the names each descendant may have.

Reviewer 2: Probably need a different example, as `payload..name` vs. `payload..*name` return the same output when given the input below. Could we make the point and show different results using something like `payload..user` and `payload..*user` on a different input, one that adds a second set of `user` elements, for example 
```
<users>
  <user>
    <name>Weave</name>
    <user>
      <name>BAT</name>
      <user>
        <name>BDD</name>
      </user>
    </user>
  </user>
  <user>
    <name>Weave2</name>
    <user>
      <name>BAT2</name>
      <user>
        <name>BDD2</name>
      </user>
    </user>
  </user>
</users>
```

Reviewer 2: In this case, `payload..user`  returns: 
```
[
  {
    "name": "Weave",
    "user": {
      "name": "BAT",
      "user": {
        "name": "BDD"
      }
    }
  },
  {
    "name": "BAT",
    "user": {
      "name": "BDD"
    }
  },
  {
    "name": "BDD"
  },
  {
    "name": "BAT2",
    "user": {
      "name": "BDD2"
    }
  },
  {
    "name": "BDD2"
  }
]
```

Reviewer 2: And `payload..*user` returns this (which includes the `user` element with the "Weave2" value):  
```
[
  {
    "name": "Weave",
    "user": {
      "name": "BAT",
      "user": {
        "name": "BDD"
      }
    }
  },
  {
    "name": "Weave2",
    "user": {
      "name": "BAT2",
      "user": {
        "name": "BDD2"
      }
    }
  },
  {
    "name": "BAT",
    "user": {
      "name": "BDD"
    }
  },
  {
    "name": "BDD"
  },
  {
    "name": "BAT2",
    "user": {
      "name": "BDD2"
    }
  },
  {
    "name": "BDD2"
  }
]
```

==== Input

[source,xml, linenums]
----

<users>
    <user>
        <name>Weave</name>
        <user>
            <name>BAT</name>
            <user>
                <name>BDD</name>
            </user>
        </user>
    </user>
</users>
----

==== Source

[source,DataWeave, linenums]
----
%dw 2.0
output application/json
---
payload..
----

==== Output

[source,json,linenums]
----
[
  {
    "user": {
      "name": "Weave",
      "user": {
        "name": "BAT",
        "user": {
          "name": "BDD"
        }
      }
    }
  },
  {
    "name": "Weave",
    "user": {
      "name": "BAT",
      "user": {
        "name": "BDD"
      }
    }
  },
  "Weave",
  {
    "name": "BAT",
    "user": {
      "name": "BDD"
    }
  },
  "BAT",
  {
    "name": "BDD"
  },
  "BDD"
]
----

=== Example

Shows how to select only descendants names

==== Input

[source,xml, linenums]
----

<users>
    <user>
        <name>Weave</name>
        <user>
            <name>BAT</name>
            <user>
                <name>BDD</name>
            </user>
        </user>
    </user>
</users>
----

==== Source

[source,DataWeave, linenums]
----
%dw 2.0
output application/json
---

> payload..name

Reviewer 2: Note that `payload..name` returns the same output as `payload..*name`.
